### PR TITLE
Sd sync save mobile sessions data

### PR DESCRIPTION
**Pull request by anna1901**

### CODE REVIEW

File: AirCasting/ABConnector/SDCardAirBeamServices.swift [hunk 1]
```diff
@@ -51,14 +47,17 @@ class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
...
+        var currentSessionTypeExpected: Int = 0
         metadataCharacteristicObserver = bluetoothManager.subscribeToCharacteristic(DOWNLOAD_META_DATA_FROM_SD_CARD_CHARACTERISTIC_UUID) { result in
             switch result {
             case .success(let data):
                 guard let data = data, let payload = String(data: data, encoding: .utf8) else {
                     self.finishSync { completion(.failure(SDCardSyncError.cantDecodePayload)) }
                     return
                 }
-                self.currentSessionType = self.currentSessionType.next
+                currentSessionType = currentSessionType.next
```
Comment (jankrzempek) [inline AirCasting/ABConnector/SDCardAirBeamServices.swift:60]: Guard maybe not to unwrap it below then?

File: AirCasting/SDSync/SDCardMeasurementsParser.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            Log.warning("Wrong data format in the csv row: \(lineSting)")
+            return nil
+        }
+        return SDCardMeasurementsRow(sessionUUID: sessionUUID, date: date, lat: lat, long: long, f: f, rh: rh, pm1: pm1, pm2_5: pm2_5, pm10: pm10)
+    }
+    
+    private func dateFrom(date: Substring, time: Substring) -> Date? {
+        let isoDate = String(date + "T" + time)
+        let dateFormatter = DateFormatters.SDCardSync.fileParserFormatter
+        let date = dateFormatter.date(from:isoDate)
```
Comment (jankrzempek) [inline AirCasting/SDSync/SDCardMeasurementsParser.swift:45]: Can we please use space after [from:]? 😄

File: AirCasting/SDSync/SDCardMobileSessionsSavingService.swift [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        
+        measurementStreamStorage.accessStorage { storage in
+            do {
+                try self.fileLineReader.readLines(of: fileURL, progress: { line in
+                    switch line {
+                    case .line(let content):
+                        let measurementsRow = self.parser.parseMeasurement(lineSting: content)
+                        guard let measurements = measurementsRow, !sessionsToIgnore.contains(measurements.sessionUUID) else { return }
+                        
+                        var session = processedSessions.first(where: {$0.uuid == measurements.sessionUUID })
```
Comment (jankrzempek) [inline AirCasting/SDSync/SDCardMobileSessionsSavingService.swift:39]: Can we please space between { and $0

File: AirCasting/SDSync/SDCardMobileSessionsSavingService.swift [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                    Log.info("Error setting status to finished and updating end time: \(error)")
+                }
+            }
+        }
+    }
+    
+    private func createSession(storage: HiddenCoreDataMeasurementStreamStorage, sdStream: SDStream, location: CLLocationCoordinate2D?, time: Date?, sessionsToCreate: inout [SessionUUID]) {
+        do {
+            try storage.createSession(Session(uuid: sdStream.sessionUUID, type: .mobile, name: "Imported from SD card", deviceType: .AIRBEAM3, location: location, startTime: time))
+            sessionsToCreate.removeAll(where: {$0 == sdStream.sessionUUID })
```
Comment (jankrzempek) [inline AirCasting/SDSync/SDCardMobileSessionsSavingService.swift:99]: Can we please space between { and $0

File: AirCasting/SDSync/SDCardMobileSessionsSavingService.swift [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    }
+    
+    private func processSession(storage: HiddenCoreDataMeasurementStreamStorage, sessionUUID: SessionUUID, deviceID: String, sessionsToIgnore: inout [SessionUUID], sessionsToCreate: inout [SessionUUID]) -> SDSession? {
+        if let existingSession = try? storage.getExistingSession(with: sessionUUID) {
+            guard existingSession.isInStandaloneMode && existingSession.sensorPackageName == deviceID else {
+                Log.info("## Ignoring session \(existingSession.name ?? "none")")
+                sessionsToIgnore.append(sessionUUID)
+                return nil
+            }
+
```
Comment (jankrzempek) [inline AirCasting/SDSync/SDCardMobileSessionsSavingService.swift:125]: Do we want this space 😄?

**Review by jankrzempek — COMMENTED** [review]
Amazing stuff.

File: AirCasting/ABConnector/BluetoothManager.swift [hunk 6]
```diff
@@ -148,7 +148,6 @@ extension BluetoothManager: CBCentralManagerDelegate {
     func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {
         if !devices.contains(peripheral) {
             if peripheral.name != nil {
-                guard !mobilePeripheralSessionManager.standaloneSessionInProgressWith(peripheral) else { return }
```
Comment (mturnau) [inline AirCasting/ABConnector/BluetoothManager.swift:151]: I need an explenation here :)

File: AirCasting/ABConnector/SDCardAirBeamServices.swift [hunk 7]
```diff
@@ -74,8 +73,8 @@ class BluetoothSDCardAirBeamServices: SDCardAirBeamServices {
                           self.finishSync { completion(.failure(SDCardSyncError.unexpectedMetadataFormat)) }
                           return
                       }
-                self.currentSessionTypeExpected = measurementsCount
-                progress(.metadata(SDCardMetaData(sessionType: self.currentSessionType!, measurementsCount: measurementsCount)))
+                currentSessionTypeExpected = measurementsCount
```
Comment (mturnau) [inline AirCasting/ABConnector/SDCardAirBeamServices.swift:76]: couple topics to discuss:
- if we encounter unexpected format, we should rather 0 the counter (that's what we do in Android
- structure the code so it's more clear what step of sync this is
- we also get meta data once the card is cleared (SD_DELETE_FINISH) and I think we are not handling it here?

File: AirCasting/ABConnector/SDSyncController.swift [hunk 8]
```diff
@@ -38,16 +46,20 @@ class SDSyncController: ObservableObject {
             self.writingQueue.sync {
                 switch result {
                 case .success:
-                    self.checkFilesForCorruption()
+                    let files = self.fileWriter.finishAndSave()
+                    self.checkFilesForCorruption(files)
```
Comment (mturnau) [inline AirCasting/ABConnector/SDSyncController.swift:50]: what happens if data is corrupted?

File: AirCasting/CoreData/MeasurementStreamStorage.swift [hunk 9]
```diff
@@ -165,6 +180,13 @@ final class HiddenCoreDataMeasurementStreamStorage: MeasurementStreamStorageCont
 
         return newStream.localID
     }
+    
+    func setStatusToFinishedAndUpdateEndTime(for sessionUUID: SessionUUID, endTime: Date?) throws {
```
Comment (mturnau) [inline AirCasting/CoreData/MeasurementStreamStorage.swift:184]: not sure if this belongs to measurementStreamstorage

File: AirCasting/SDSync/SDCardMobileSessionsSavingService.swift [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        }
+    }
+    
+    private func read(fileURL: URL) -> PassthroughSubject<SDCardMeasurementsRow?, Error> {
+        let publisher = PassthroughSubject<SDCardMeasurementsRow?, Error>()
+        do {
+            try fileLineReader.readLines(of: fileURL, progress: { line in
+                switch line {
+                case .line(let content):
+                    let measurementsRow = parser.parseMeasurement(lineSting: content)
```
Comment (mturnau) [inline AirCasting/SDSync/SDCardMobileSessionsSavingService.swift:66]: was that supposed to be `lineString`?

File: AirCasting/SDSync/SDCardMobileSessionsSavingService.swift [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        streamsWithMeasurements[SDStream(sessionUUID: measurements.sessionUUID, name: .pm1), default: []].append(Measurement(time: measurements.date, value: measurements.pm1, location: CLLocationCoordinate2D(latitude: measurements.lat, longitude: measurements.long)))
+        streamsWithMeasurements[SDStream(sessionUUID: measurements.sessionUUID, name: .pm2_5), default: []].append(Measurement(time: measurements.date, value: measurements.pm2_5, location: CLLocationCoordinate2D(latitude: measurements.lat, longitude: measurements.long)))
+        streamsWithMeasurements[SDStream(sessionUUID: measurements.sessionUUID, name: .pm10), default: []].append(Measurement(time: measurements.date, value: measurements.pm10, location: CLLocationCoordinate2D(latitude: measurements.lat, longitude: measurements.long)))
+    }
+    
+    private func createMeasurementStream(for sensorName: StreamSensorName, sensorPackageName: String) -> MeasurementStream {
+        switch sensorName {
```
Comment (mturnau) [inline AirCasting/SDSync/SDCardMobileSessionsSavingService.swift:142]: hmm

**Review by mturnau — APPROVED** [review]

**Pull request merged**